src/box/lua/config/applier/sharding.lua

```diff
@@ -21,6 +21,17 @@ local function apply(config)
     end
     local cfg = configdata:sharding()
     if is_storage then
+        local function deploy_funcs()
+            local vexports = require('vshard.storage.exports')
+            local exports = vexports.compile(vexports.log[#vexports.log])
+            vexports.deploy_funcs(exports)
+        end
+        box.watch('box.status', function(_, status)
+            -- It's ok, if deploy_funcs() will be triggered several times.
+            if status.is_ro == false then
+                deploy_funcs()
+            end
+        end)
```

Comment:
1. A new watcher will start at each reload.
2. Shouldn't we re-check that the instance has the `storage` sharding role on each RO -> RW transition (and stop the watcher if we're not storage anymore)?

Reply:
1. True, fixed.
2. For now I'm only checking if the instance has the storage sharding role on startup and reload. The watcher now stops if the storage sharding role has been removed. The storage sharding role shouldn't be lost on RO->RW transition, or is there a way?

Reply:
Thanks!

> The storage sharding role shouldn't be lost on RO->RW transition, or is there a way?

I see no such a way.